src/adapters/tcp.rs

```diff
@@ -102,7 +123,25 @@ impl Remote for RemoteResource {
     }
 
     fn pending(&self, _readiness: Readiness) -> PendingStatus {
-        check_stream_ready(&self.stream)
+        let status = check_stream_ready(&self.stream);
+
+        if status == PendingStatus::Ready {
+            if let Some(keepalive) = &self.keepalive {
+                #[cfg(target_os = "windows")]
+                let socket = unsafe { Socket::from_raw_socket(self.stream.as_raw_socket()) };
+                #[cfg(not(target_os = "windows"))]
+                let socket = unsafe { Socket::from_raw_fd(self.stream.as_raw_fd()) };
+
+                if let Err(e) = socket.set_tcp_keepalive(keepalive) {
+                    log::warn!("TCP set keepalive error: {}", e);
+                }
+
+                // Don't drop so the underlying socket is not closed.
+                forget(socket);
+            }
+        }
+
+        status
```

Comment:
I'm wondering why do this here instead of in `connect()`/`listen()` methods.

Could we get rid of the `unsafe` code building the socket with `socket2`, and once it's built, get the `stream`? instead of building the stream and getting the socket from it?

Reply:
We cannot do that in `listen()` because it concerns the accepted socket, not the listener socket. We could do it in `accept()` but `mio` does quite a bit of platform dependent stuff [in the UNIX path](https://docs.rs/mio/latest/x86_64-unknown-linux-gnu/src/mio/sys/unix/tcp.rs.html#54) which contains unsafe as well. We would loose that or have to copy and maintain it ourselves. I'd rather avoid that and live with the unsafe code.

 ([The Windows path](https://docs.rs/mio/latest/x86_64-pc-windows-msvc/src/mio/sys/windows/tcp.rs.html) just calls `std::net::TcpListener::accept()`, though.)

Reply:
Ok, thanks for your thoughts!